subscription_test.go

```diff
@@ -0,0 +1,196 @@
+package graphql_test
+
+import (
+	"context"
+	"encoding/json"
+	stdErrors "errors"
+	"testing"
+
+	graphql "github.com/graph-gophers/graphql-go"
+	"github.com/graph-gophers/graphql-go/errors"
+	"github.com/graph-gophers/graphql-go/gqltesting"
+)
+
+type rootResolver struct {
+	*helloResolver
+	*helloSaidResolver
+}
+
+type helloResolver struct{}
+
+func (r *helloResolver) Hello() string {
+	return "Hello world!"
+}
+
+var resolverErr = stdErrors.New("resolver error")
+
+type helloSaidResolver struct {
+	err      error
+	upstream <-chan *helloSaidEventResolver
+}
+
+type helloSaidEventResolver struct {
+	msg string
+	err error
+}
+
+func (r *helloSaidResolver) HelloSaid(ctx context.Context) (chan *helloSaidEventResolver, error) {
+	if r.err != nil {
+		return nil, r.err
+	}
+
+	c := make(chan *helloSaidEventResolver)
+	go func() {
+		for r := range r.upstream {
+			select {
+			case <-ctx.Done():
+				close(c)
+				return
+			case c <- r:
+			}
+		}
+		close(c)
+	}()
+
+	return c, nil
+}
+
+func (r *helloSaidEventResolver) Msg() (string, error) {
+	return r.msg, r.err
+}
+
+func closedUpstream(rr ...*helloSaidEventResolver) <-chan *helloSaidEventResolver {
+	c := make(chan *helloSaidEventResolver, len(rr))
+	for _, r := range rr {
+		c <- r
+	}
+	close(c)
+	return c
+}
+
+func TestSchemaSubscribe(t *testing.T) {
+	gqltesting.RunSubscribes(t, []*gqltesting.TestSubscription{
+		{
+			Name: "ok",
+			Schema: graphql.MustParseSchema(schema, &rootResolver{
+				helloSaidResolver: &helloSaidResolver{
+					upstream: closedUpstream(
+						&helloSaidEventResolver{msg: "Hello world!"},
+						&helloSaidEventResolver{err: resolverErr},
+						&helloSaidEventResolver{msg: "Hello again!"},
+					),
+				},
+			}),
+			Query: `
+				subscription onHelloSaid {
+					helloSaid {
+            msg
+          }
+				}
+			`,
+			ExpectedResults: []gqltesting.TestResponse{
+				{
+					Data: json.RawMessage(`
+						{
+							"helloSaid": {
+								"msg": "Hello world!"
+							}
+						}
+					`),
+				},
+				{
+					Data: json.RawMessage(`
+						{
+							"helloSaid": {
+								"msg":null
+							}
+						}
+					`),
+					Errors: []*errors.QueryError{errors.Errorf("%s", resolverErr)},
+				},
+				{
+					Data: json.RawMessage(`
+						{
+							"helloSaid": {
+								"msg": "Hello again!"
+							}
+						}
+					`),
+				},
+			},
+		},
+		{
+			Name:   "parse_errors",
+			Schema: graphql.MustParseSchema(schema, &rootResolver{}),
+			Query:  `invalid graphQL query`,
+			ExpectedResults: []gqltesting.TestResponse{
+				{
+					Errors: []*errors.QueryError{errors.Errorf("%s", `syntax error: unexpected "invalid", expecting "fragment" (line 1, column 9)`)},
+				},
+			},
+		},
+		{
+			Name:   "subscribe_to_query_errors",
```

Comment:
Is it ever possible to get into this case in an actual application? The GraphQL lib will execute a Query if the operation is `query`, no? I don't see the reason for this test. Am I missing something!

Reply:
This patch exposes `schema.Subscribe(ctx context.Context, queryString string, operationName string, variables map[string]interface{}) (<-chan *Response, error)` which will be leveraged by the transport. The queryString could contain anything. I'm not sure if it makes sense to leave this job to the transport layer.